ui/apps/platform/src/Containers/NetworkGraph/utils/networkGraphUtils.ts

```diff
@@ -0,0 +1,64 @@
+import { Model, NodeModel } from '@patternfly/react-topology';
+
+import { ListenPort } from 'types/networkFlow.proto';
+
+/* node helper functions */
+
+function getExternalNodeIds(model: Model): string[] {
+    const externalNodeIds =
+        model.nodes?.reduce((acc, curr) => {
+            if (curr.data.type === 'INTERNET' || curr.data.type === 'EXTERNAL_SOURCE') {
+                return [...acc, curr.id];
+            }
+            return acc;
+        }, [] as string[]) || [];
+    return externalNodeIds;
+}
+
+export function getNodeById(model: Model, nodeId: string): NodeModel | undefined {
+    return model.nodes?.find((node) => node.id === nodeId);
+}
+
+/* edge helper functions */
+
+export function getNumInternalFlows(model: Model, deploymentId: string): number {
+    const externalNodeIds = getExternalNodeIds(model);
+    const numInternalFlows =
+        model.edges?.reduce((acc, edge) => {
+            if (
+                (edge.source === deploymentId && !externalNodeIds.includes(edge.target || '')) ||
+                (edge.target === deploymentId && !externalNodeIds.includes(edge.source || ''))
+            ) {
+                return acc + 1;
+            }
+            return acc;
+        }, 0) || 0;
+    return numInternalFlows;
+}
+
+export function getNumExternalFlows(model: Model, deploymentId: string): number {
+    const externalNodeIds = getExternalNodeIds(model);
+    const numExternalFlows =
+        model.edges?.reduce((acc, edge) => {
+            if (
+                (edge.source === deploymentId && externalNodeIds.includes(edge.target || '')) ||
+                (edge.target === deploymentId && externalNodeIds.includes(edge.source || ''))
+            ) {
+                return acc + 1;
+            }
+            return acc;
+        }, 0) || 0;
+    return numExternalFlows;
+}
+
+/* deployment helper functions */
+
+export function getListenPorts(model: Model, deploymentId: string): ListenPort[] {
+    const deployment = model.nodes?.find((node) => {
+        return node.id === deploymentId;
+    });
```

Comment:
i think you can use `controller.getNodeById(id)` here too

Reply:
Ah but then I'd have to pass the `controller` or use the hook (`useVisualizationController`). I'd have to rethink the flow of data here. I might have to have the `DeploymentSideBar` use the `controller` and prepare the data to be passed to `DeploymentDetails` in a different way. Are there more helper functions that the `controller` offers?

Reply:
thats true. we should see if it makes more or less sense to pass the controller like that around. somewhat, you can check the API doc i made (https://docs.google.com/document/d/1j0P1cqOJ69VWDOlpemSejlkL_ya99suSp1IvgqojeGA/edit#heading=h.v6s7p22e33i8) -- there are also additional functions for each node/edge element defined in the library too